Add support for alarm system, switch and thermostat to homekit

maxclaey opened the pull request:
## Description:

Add support for alarm system, thermostat and switch (and remote represented by a switch) to homekit.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) (if applicable):** home-assistant/home-assistant.github.io#4838

## Checklist:
  - [x] The code change is tested and works locally.

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

maxclaey committed:
Added support for security system, switch and thermostat

maxclaey committed:
Fix too long lines

maxclaey committed:
Fix too long lines

maxclaey committed:
Fix too long lines

cdce8p commented on homeassistant/components/homekit/thermostats.py:
```diff
@@ -0,0 +1,135 @@
+                                     {'entity_id': self._entity_id,
+                                      'operation_mode': hass_value})
+
+    def set_temperature_debounced(self, value):
```
Is this necessary?

maxclaey replied:
Necessary is a big word off course :). But I noticed for myself that if you use the Home app to set the temperature, sliding the temperature from 21 to 16 degrees, causes set temperature calls to 20, 19, 18, 17 and 16, while all I want to do is set the temperature to 16 degrees. To avoid sending to much set calls to the device (Nest thermostat in my case), I though debouncing was a good idea.

cdce8p replied:
I get the idea, but you can select a discrete value in HomeKit. Which only results in one call to the method anyway.

cdce8p reviewed (changes requested):
Great first PR. Looking forward to adding the support for all these components. Just added a few comments.

maxclaey committed:
Processing review of @cdce8p (greatly appreciated!)

maxclaey commented on homeassistant/components/homekit/switches.py:
```diff
@@ -0,0 +1,63 @@
+        """Update switch state after state changed."""
+        if new_state is None:
+            return
+
```
Probably I'm missing something, but how can I know if the call came from Home Assistant or not? Furthermore, as far as I can see, update_state is only called from Switch.run and through the handler of the track state change, so it can only be called through Home Assistant, right?

cdce8p replied:
No quite. A `state_changed` event is also called if HomeKit wants to set the state value for Home Assistant. There a two cases in which `update_state` is called:

* `HomeKit value change` -> `HA state_change` -> `update_state`
* `HA value change` -> `HA state_change` -> `update_state`

During my first review I suggested to use `should_callback=False`. This makes sure that `set_state` is only called from HomeKit. Sadly there is no such option for the Home Assistant side. The best solution I have come up with so far is to set a flag var in `set_state` and check/remove it during `update_state`.

maxclaey replied:
Ok, I see. I guess the only "problem" is performing the set_value on the characteristic? I now only do it when the call did not come from HomeKit (i.e. when homekit_target_on is None). The remainder of the update_state is needed in my opinion to have current_on set correctly

maxclaey commented on homeassistant/components/homekit/security_systems.py:
```diff
@@ -0,0 +1,88 @@
+            service = STATE_TO_SERVICE[hass_value]
+
+            self._hass.services.call('alarm_control_panel', service,
+                                     {'entity_id': self._entity_id})
```
Hmm, I will have to think about this :). In my case, I have an alarm without code in home assistant and I have to admit I have overseen this loss of generality. I will try to have a look at it ASAP

cdce8p replied:
I don't have much time at the moment, but maybe plan that the alarm code will be part of the constructor with: alarm_code=None. I will explain later.

cdce8p replied:
The current implementation for HomeKit has a huge bug. If you have multiple accessories, it might happen that after a restart of Home Assistant some stop working an only show `Unavailable`. This is due to the way pyhap generates accessory ids. If after a restart those change they will not work.
I plan to add a config option this weekend so that the aid can be set/stored and doesn't change anymore. It would be quiet easy to add an additional parameter for `alarm_panels` which could be used to initialize the accessory. Until then it will only work if someone doesn't have a alarm_code set.

maxclaey replied:
I added the alarm_code parameter defaulting to None to the constructor as requested. When it is set, it is passed in the service call to change the alarm state.
Regarding the HomeKit bug: I saw it happening before that accessories stop working after reboot (blocked in the home app), so this explains why..

maxclaey commented on homeassistant/components/homekit/security_systems.py:
```diff
@@ -0,0 +1,88 @@
+        self.current_security_state = None
+        self.homekit_target_security_state = None
+
+        self.service_alarm = self.get_service(SERV_SECURITY_SYSTEM)
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/security_systems.py:
```diff
@@ -0,0 +1,88 @@
+from homeassistant.helpers.event import async_track_state_change
+
+from . import TYPES
+from .accessories import HomeAccessory
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/security_systems.py:
```diff
@@ -0,0 +1,88 @@
+        self.homekit_target_security_state = None
+
+        self.service_alarm = self.get_service(SERV_SECURITY_SYSTEM)
+        self.char_current_state = self.service_alarm.\
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/security_systems.py:
```diff
@@ -0,0 +1,88 @@
+        self.service_alarm = self.get_service(SERV_SECURITY_SYSTEM)
+        self.char_current_state = self.service_alarm.\
+            get_characteristic(CHAR_CURRENT_SECURITY_STATE)
+        self.char_target_state = self.service_alarm.\
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/security_systems.py:
```diff
@@ -0,0 +1,88 @@
+from .accessories import HomeAccessory
+from .const import (SERV_SECURITY_SYSTEM, CHAR_CURRENT_SECURITY_STATE,
+                    CHAR_TARGET_SECURITY_STATE)
+from homeassistant.const import (STATE_ALARM_ARMED_AWAY,
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/security_systems.py:
```diff
@@ -0,0 +1,88 @@
+                                 self.update_security_state)
+
+    def set_security_state(self, value):
+        """Move security state to value if call came from homekit."""
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/security_systems.py:
```diff
@@ -0,0 +1,88 @@
+            if self.homekit_target_security_state is None or \
+                    self.homekit_target_security_state \
+                    == self.current_security_state:
+                self.char_target_state.set_value(self.current_security_state)
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/switches.py:
```diff
@@ -0,0 +1,63 @@
+from homeassistant.helpers.event import async_track_state_change
+
+from . import TYPES
+from .accessories import HomeAccessory
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/switches.py:
```diff
@@ -0,0 +1,63 @@
+        self.current_on = None
+        self.homekit_target_on = None
+
+        self.service_switch = self.get_service(SERV_SWITCH)
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/switches.py:
```diff
@@ -0,0 +1,63 @@
+                                 self.update_state)
+
+    def set_state(self, value):
+        """Move switch state to value if call came from homekit."""
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/switches.py:
```diff
@@ -0,0 +1,63 @@
+        self.current_on = (new_state.state == 'on')
+        self.char_on.set_value(self.current_on)
+        _LOGGER.debug("%s: Updated current state to %s (%s)",
+                      self._entity_id, new_state.state, self.current_on)
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/switches.py:
```diff
@@ -0,0 +1,63 @@
+        _LOGGER.debug("%s: Want to update current state to %s",
+                      self._entity_id, new_state.state)
+        self.current_on = (new_state.state == 'on')
+        self.char_on.set_value(self.current_on)
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/switches.py:
```diff
@@ -0,0 +1,63 @@
+
+        self._hass = hass
+        self._entity_id = entity_id
+        self._domain = hass.states.get(self._entity_id).domain
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/switches.py:
```diff
@@ -0,0 +1,63 @@
+            return
+
+        _LOGGER.debug("%s: Want to update current state to %s",
+                      self._entity_id, new_state.state)
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -89,6 +95,19 @@ def get_accessory(hass, state):
+        _LOGGER.debug("Add \"%s\" as \"%s\"", state.entity_id, 'Thermostat')
+        return TYPES['Thermostat'](hass, state.entity_id, state.name)
+
+    elif state.domain == 'switch' or state.domain == 'remote':
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/thermostats.py:
```diff
@@ -0,0 +1,135 @@
+from homeassistant.helpers.event import async_track_state_change
+
+from . import TYPES
+from .accessories import HomeAccessory
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/thermostats.py:
```diff
@@ -0,0 +1,135 @@
+        self.target_temperature = None
+        self.display_units = None
+
+        self.service_thermostat = self.get_service(SERV_THERMOSTAT)
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/thermostats.py:
```diff
@@ -0,0 +1,135 @@
+
+        if self.homekit_target_heat_cool is None \
+                or self.homekit_target_heat_cool == self.current_heat_cool:
+            self.char_target_heat_cool.set_value(self.current_heat_cool)
```
Thanks, done!

maxclaey commented on homeassistant/components/homekit/thermostats.py:
```diff
@@ -0,0 +1,135 @@
+                                 self.update_thermostat)
+
+    def set_heat_cool(self, value):
+        """Move operation mode to value if call came from homekit."""
```
Thanks, done!

cdce8p commented:
Could you rebase your branch againt `upstream/dev` to resolve the conflicts? https://home-assistant.io/developers/development_catching_up/

maxclaey committed:
Only perform set call when the call didn't come from HomeKit

maxclaey committed:
Added support for alarm_code

cdce8p commented on homeassistant/components/homekit/security_systems.py:
```diff
@@ -0,0 +1,92 @@
+            return
+
+        hass_state = new_state.state
+        _LOGGER.debug("%s: Want to update current state to %s",
```
Remove the logger message.

cdce8p commented on homeassistant/components/homekit/security_systems.py:
```diff
@@ -0,0 +1,92 @@
+                      self._entity_id, hass_state)
+        if hass_state in HASS_TO_HOMEKIT:
+            self.current_security_state = HASS_TO_HOMEKIT[hass_state]
+            self.char_current_state.set_value(self.current_security_state)
```
```py
if hass_state not in HASS_TO_HOMEKIT:
    return
current_security_state = HASS_TO_HOMEKIT[hass_state]
self.char_current_state.set_value(current_security_state)
...
...
```

cdce8p commented on homeassistant/components/homekit/security_systems.py:
```diff
@@ -0,0 +1,92 @@
+            self.char_current_state.set_value(self.current_security_state)
+            _LOGGER.debug("%s: Updated current state to %s (%d)",
+                          self._entity_id, hass_state,
+                          self.current_security_state)
```
`current_security_state`

cdce8p commented on homeassistant/components/homekit/security_systems.py:
```diff
@@ -0,0 +1,92 @@
+                    == self.current_security_state:
+                self.char_target_state.set_value(self.current_security_state,
+                                                 should_callback=False)
+                self.homekit_target_security_state = None
```
The whole if statement can be improved. The following should work:
```py
if not self.flag_target_state:
    self.char_target_state.set_value(current_security_state, should_callback=False)
elif self.char_target_state.get_value() == self.char_current_state.get_value():
    self.flag_target_state = False
```
* `if not self.flag_target_state` will set the target state if change came from Home Assistant.
* `elif self.char_target_state.get_value() == self.char_current_state.get_value()` will reset flag var if change came from HomeKit.

cdce8p reviewed (changes requested):
Added a few comments to `init`, `const` and `security_systems`. Will review the other files next.

cdce8p commented on homeassistant/components/homekit/thermostats.py:
```diff
@@ -0,0 +1,137 @@
+        if (self.homekit_target_heat_cool is None
+            and self.current_heat_cool is not None) \
+                or self.homekit_target_heat_cool == self.current_heat_cool:
+            self.char_target_heat_cool.set_value(self.current_heat_cool,
```
What if `self.current_heat_cool = None`? Should result in an value error at the moment.

cdce8p commented on homeassistant/components/homekit/thermostats.py:
```diff
@@ -0,0 +1,137 @@
+            self._hass.services.call('climate', 'set_temperature',
+                                     {'entity_id': self._entity_id,
+                                      'temperature':
+                                          self.homekit_target_temperature})
```
```py
self._hass.services.call(
    'climate', 'set_temperature', {
        ATTR_ENTITIY_ID: self._entity_id,
        ATTR_TEMPERATURE: value})
```

cdce8p commented on homeassistant/components/homekit/thermostats.py:
```diff
@@ -0,0 +1,137 @@
+                                          self.homekit_target_temperature})
+
+    def update_thermostat(self, entity_id=None,
+                          old_state=None, new_state=None):
```
I haven't look at this method in detail, but it might be a good idea to use flag vars as well, like I suggested above.

cdce8p commented on homeassistant/components/homekit/thermostats.py:
```diff
@@ -0,0 +1,137 @@
+            self.char_display_units.set_value(self.display_units)
+
+        if (self.homekit_target_heat_cool is None
+            and self.current_heat_cool is not None) \
```
Lint: Add 4 spaces

cdce8p reviewed (commented):
Some more suggestions.

cdce8p commented:
Could you add tests? Take a look at the ones that are already their https://github.com/home-assistant/home-assistant/tree/dev/tests/components/homekit for help.

cdce8p commented:
Add:
```py
'homeassistant.components.homekit.security_systems',
'homeassistant.components.homekit.switches',
'homeassistant.components.homekit.thermostats',
```
to https://github.com/home-assistant/home-assistant/blob/dev/script/gen_requirements_all.py#L94-L99 to prevent pylint errors.

cdce8p commented:
Add:
```py
'homeassistant.components.homekit.security_systems',
'homeassistant.components.homekit.switches',
'homeassistant.components.homekit.thermostats',
```
to https://github.com/home-assistant/home-assistant/blob/dev/script/gen_requirements_all.py#L94-L99 to prevent pylint errors.

maxclaey committed:
Take into account review remarks

maxclaey committed:
Provide tests for HomeKit security systems, switches and thermostats

maxclaey committed:
Fix hound warnings

cdce8p commented on homeassistant/components/homekit/thermostats.py:
```diff
@@ -0,0 +1,137 @@
+        """Update security state after state changed."""
+        if new_state is None:
+            return
+        current_temp = new_state.attributes[ATTR_CURRENT_TEMPERATURE]
```
Add an empty line to improve readability. Also between currently L118/L119 and L127/L128.

cdce8p commented on homeassistant/components/homekit/thermostats.py:
```diff
@@ -0,0 +1,137 @@
+        self.char_target_temp = self.service_thermostat. \
+            get_characteristic(CHAR_TARGET_TEMPERATURE)
+        self.char_target_temp.value = 0.0
+        self.char_target_temp.setter_callback = self.set_temperature_debounced
```
Use `self.set_target_temperature` instead. We shouldn't use `self.set_temperature_debounced` as I've explained.

maxclaey replied:
I removed the debounced version based on your request, even though I still think it is needed. When you use the standard iOS Home app, you have to use a slider to set the temperature, resulting in a lot of unnecessary calls.

cdce8p replied:
It arent't that many more. In most cases it's infect only one: like automations or using Siri. Even if you use the slider I hardly get more then two calls. If you use the step increment in Home Assistant, it's more.
Until a moment ago I just had a feeling not to use it since `timer` is imported from `threading` which usually means that it is somehow blocking something else. I looked it up and `timer` initialize a new thread with each init call. In Home Assistant we try to use async as much as possible to improve performance even on something like a Raspberry Pi. That isn't to say that the timer function couldn't be build using sth like `asyncio`, but IMO it is not worth it at the moment. There are bigger projects to work on, like support for Z-Wave device (which sadly will require some major changes to all `homekit` files) or making the `pyhap` libary asynchronous.

cdce8p commented on homeassistant/components/homekit/thermostats.py:
```diff
@@ -0,0 +1,137 @@
+                self.char_target_heat_cool.set_value(
+                    HC_HASS_TO_HOMEKIT[operation_mode],
+                    should_callback=False)
+        display_units = new_state.attributes['unit_of_measurement']
```
Use `ATTR_UNIT_OF_MEASUREMENT` imported from `homeassistant.const` instead.

cdce8p commented:
Some last changes. `security_systems` and `switches` are looking quite good. For `thermostats`: it seems like `STATE_AUTO` and temperature ranges doesn't work jet? That can be improved later.
The last thing that is be needed would be an update to the documentation against branch `next`: https://github.com/home-assistant/home-assistant.github.io/blob/next/source/_components/homekit.markdown#-linkable_title-supported-components-

maxclaey committed:
Support STATE_AUTO

maxclaey commented:
I initially only added support for the required HomeKit thermostat characteristics. I now added support for `STATE_AUTO` and temperature ranges as well. The problem there however is that there is a mismatch between HomeKit and home assistant here. Home assistant can have 'off', 'heat', 'cool' and 'auto' both as current and as target mode. However, for HomeKit, you can have 'auto' as target mode, but current mode can only be 'off', 'heat' and 'cool', so we have to do some dirty stuff when the home assistant state is set to 'auto'. Please note that I could not test this feature as my nest device does not support it

maxclaey committed:
Guard if state exists

maxclaey committed:
Fix lint error

cdce8p commented:
For testing use the `demo` [platforms](https://home-assistant.io/components/demo/). Works for many other components similarly.
```py
climate:
  - platform: demo
```

cdce8p commented on homeassistant/components/homekit/thermostats.py:
```diff
@@ -0,0 +1,214 @@
+                self.char_cooling_thresh_temp.set_value(cooling_thresh_temp,
+                                                        should_callback=False)
+
+        # Update heating theshold temperature
```
Typo `threshold`

cdce8p commented on homeassistant/components/homekit/thermostats.py:
```diff
@@ -0,0 +1,214 @@
+        if state is not None:
+            supported_features = state.attributes[ATTR_SUPPORTED_FEATURES]
+            if supported_features is not None and \
+                    supported_features & SUPPORT_TARGET_TEMPERATURE_HIGH:
```
This should be done in `__init__.get_accessory` and it should be a new type (one that supports thresholds).

maxclaey replied:
Ok, I will do the feature support check in `__init__.get_accessory`. What do you mean with a new type? Can I just pass a constructor parameter to Thermostat saying if it supports this feature or not? Or should I make a "new" type that inherits from Thermostat? Or make a completely new type for it (causing a lot of code duplication obviously)? I will make all requested changes tonight (Belgian time)

cdce8p replied:
I haven't figured out the best way jet, but the issue is that as soon as you add additional characteristics to `SERV_THERMOSTAT`, HomeKit assumes that the device supports those.

I'm not sure if one of those ways would be better. The reason I suggested to test it in `get_accessory` is that this is the option I have chosen for `sensors`. Would work as well if it would be before `add_preload_service`. Choose the one that you find most convenient .

cdce8p commented on homeassistant/components/homekit/thermostats.py:
```diff
@@ -0,0 +1,214 @@
+        self.cooling_thresh_temp_flag_target_state = False
+        self.heating_thresh_temp_flag_target_state = False
+
+        self.service_thermostat = add_preload_service(self, SERV_THERMOSTAT)
```
You would need to add the optional characteristics to add to the service as a list here. Otherwise they wouldn't be initiated for the accessory.
```py
add_preload_service(self, SERV_THERMOSTAT, [CHAR_COOLING_THRESHOLD_TEMPERATURE, CHAR_HEATING_THRESHOLD_TEMPERATURE])
```

cdce8p commented:
Regarding `STATE_AUTO`: I noticed the problem, too. Thats why I suggested it could be added later. Your solution seems to be quite good though.

To finish up, could you add tests for it and in `tests/components/homekit/test_get_accessory.py` as well?

cdce8p commented:
Just saw the lint errors: https://travis-ci.org/home-assistant/home-assistant/jobs/349448789#L570

You can run the checks also locally: https://home-assistant.io/developers/development_testing/#testing-outside-of-tox after the requirements are installed using `script/lint`

maxclaey committed:
Improve support for thermostat auto mode

maxclaey committed:
Fix typo

maxclaey committed:
Provide both high and low at the same time for home assistant

maxclaey committed:
Set default values within accepted ranges

maxclaey committed:
Added tests for auto mode

maxclaey committed:
Fix hound error

maxclaey commented:
I added tests for the auto mode, as requested. However, tests are failing when I pass the additional characteristics to `add_preload_service` as you suggested (see error below). However, when I just call `self.service_thermostat = add_preload_service(self, SERV_THERMOSTAT)` (as before), everything works just fine.

The test error when passing additional parameters:
`homeassistant/components/homekit/thermostats.py:57: in __init__
    extra_chars)
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/mock.py:939: in __call__
    return _mock_self._mock_call(*args, **kwargs)
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/mock.py:1005: in _mock_call
    ret_val = effect(*args, **kwargs)
tests/mock/homekit.py:23: in mock_preload_service
    acc.add_service(service)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <homeassistant.components.homekit.thermostats.Thermostat object at 0x10e66ac88>
servs = (<[AttributeError("'str' object has no attribute 'display_name'") raised in repr()] MockService object at 0x10e66ae48>,)
s = <[AttributeError("'str' object has no attribute 'display_name'") raised in repr()] MockService object at 0x10e66ae48>, c = 'CoolingThresholdTemperature'`

cdce8p commented:
@maxclaey I messed this one up. It should work if you remove the two patch statements. Just use:
```py
acc = Thermostat(self.hass, climate, 'Climate', True)
acc.run()
```

cdce8p commented on tests/components/homekit/test_thermostats.py:
```diff
@@ -0,0 +1,183 @@
+        climate = 'climate.testclimate'
+
+        with patch(PATH_ACC, side_effect=mock_preload_service):
+            with patch(PATH_FILE, side_effect=mock_preload_service):
```
Remove L114 and L115 and the test should work.

cdce8p commented on tests/components/homekit/test_thermostats.py:
```diff
@@ -0,0 +1,183 @@
+                              ATTR_TARGET_TEMP_LOW: 20.0,
+                              ATTR_CURRENT_TEMPERATURE: 18.0,
+                              ATTR_TEMPERATURE: None,
+                              'unit_of_measurement': '°C'})
```
Use `ATTR_UNIT_OF_MEASUREMENT` instead. Also below

maxclaey committed:
Fix thermostat test error
Use attributes.get instead of indexing for safety
Avoid hardcoded attributes in tests

maxclaey commented:
Thanks for clearing this out! And thanks a lot again for all your detailed reviews! I think that I now covered all your requests. Please let me know if I have overseen anything or if you want me to change something!

cdce8p committed:
Typo

maxclaey commented:
Great, thanks! It was the first time a contributed to home-assistant, or to an open-source project in general, so I really appreciate your detailed reviews and advise!
